### PR TITLE
benchmark: printSecondSample with time series

### PR DESCRIPTION
**Pull request by gyuho**
Sample
```
Actual Latency distribution:
1457999564	69.503µs	  501
1457999565	20.887µs	  499
```

### CODE REVIEW

File: tools/benchmark/cmd/timeseries.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+		}
+		i++
+	}
+	sort.Sort(tslice)
+	return tslice
+}
+
+func (ts TimeSeries) String() string {
+	buf := new(bytes.Buffer)
+	buf.WriteString("\nActual Latency distribution:\n")
```
Comment (xiang90) [inline tools/benchmark/cmd/timeseries.go]: Sample in one second (unix latency throughput)

Comment (gyuho) [inline tools/benchmark/cmd/timeseries.go]: Just fixed. thanks.

File: tools/benchmark/cmd/timeseries.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+	throughPut int64
+}
+
+type TimeSeries []timeSeries
+
+func (t TimeSeries) Swap(i, j int)      { t[i], t[j] = t[j], t[i] }
+func (t TimeSeries) Len() int           { return len(t) }
+func (t TimeSeries) Less(i, j int) bool { return t[i].timestamp < t[j].timestamp }
+
+type timePoint struct {
```
Comment (xiang90) [inline tools/benchmark/cmd/timeseries.go]: secondPoint

File: tools/benchmark/cmd/timeseries.go [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+	latency time.Duration
```
Comment (xiang90) [inline tools/benchmark/cmd/timeseries.go]: totalLatency

File: tools/benchmark/cmd/timeseries.go [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+	count   int64
+}
+
+type timePoints struct {
```
Comment (xiang90) [inline tools/benchmark/cmd/timeseries.go]: secondPoints

File: tools/benchmark/cmd/timeseries.go [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+		}
+		i++
+	}
+	sort.Sort(tslice)
+	return tslice
+}
+
+func (ts TimeSeries) String() string {
+	buf := new(bytes.Buffer)
+	buf.WriteString("\nSample in one second (unix latency throughput):\n")
```
Comment (xiang90) [inline tools/benchmark/cmd/timeseries.go]: remove the first \n?

File: tools/benchmark/cmd/report.go [hunk 6]
```diff
@@ -142,6 +151,10 @@ func (r *report) printLatencies() {
 	}
 }
 
+func (r *report) printActualLatencies() {
```
Comment (xiang90) [inline tools/benchmark/cmd/report.go]: printSecondSample

Comment (xiang90) [inline tools/benchmark/cmd/report.go]: probably we should add a flag, only print out this when --sample is set

**Comment by xiang90** [comment]
Can you please show an example output along with the latency histogram?

File: tools/benchmark/cmd/report.go (hunk 6 above)
Comment (gyuho) [inline tools/benchmark/cmd/report.go]: Sure, will change addressing all your comments. Thanks.

**Comment by gyuho** [comment]
@xiang90 PTAL.

```

./benchmark --endpoints="..." \
	put --sample --conns=1 --clients=1 --total=5000 --val-size=3 --key-space-size=1

Summary:
  Total:        39.7764 secs.
  Slowest:      0.0261 secs.
  Fastest:      0.0049 secs.
  Average:      0.0080 secs.
  Stddev:       0.0022 secs.
  Requests/sec: 125.7028
Response time histogram:
  0.005 [1]     |
  0.007 [1562]  |∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
  0.009 [2918]  |∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
  0.011 [156]   |∎∎
  0.013 [146]   |∎∎
  0.015 [99]    |∎
  0.018 [63]    |
  0.020 [38]    |
  0.022 [14]    |
  0.024 [2]     |
  0.026 [1]     |
Latency distribution:
  10% in 0.0062 secs.
  25% in 0.0067 secs.
  50% in 0.0076 secs.
  75% in 0.0081 secs.
1458011507    56.512µs      126
  90% in 0.0094 secs.
  95% in 0.0129 secs.
  99% in 0.0178 secs.
Sample in one second (unix latency throughput):
1458011506    50.608µs       56
1458011507    56.512µs      126
1458011508    47.634µs      125
1458011509    50.025µs      125
1458011510    53.001µs      123
1458011511    50.233µs      123
1458011512    51.796µs      125
1458011513    44.244µs      127
1458011514    41.656µs      124
1458011515    45.598µs      122
1458011516    44.894µs      126
1458011517    47.258µs      123
1458011518    45.889µs      121
1458011519    58.328µs      121
1458011520    51.411µs      123
1458011521    52.527µs      125
1458011522    50.031µs      124
1458011523    48.499µs      125
1458011524    48.308µs      124
1458011525    54.474µs      126
1458011526    47.968µs      126
1458011527    45.725µs      129
1458011528    47.697µs      135
1458011529    52.077µs      126
1458011530    52.89µs      127
1458011531    51.803µs      127
1458011532    48.934µs      125
1458011533    51.982µs      127
1458011534    50.591µs      129
1458011535    53.2µs      125
1458011536    44.76µs      126
1458011537    46.163µs      126
1458011538    51.918µs      128
1458011539    46.43µs      125
1458011540    49.047µs      127
1458011541    49.653µs      126
1458011542    43.295µs      126
1458011543    44.856µs      126
1458011544    47.191µs      131
1458011545    46.726µs      126
1458011546    43.852µs       43
```

**Comment by xiang90** [comment]
why there is a sample output in between?

```
Latency distribution:
  10% in 0.0062 secs.
  25% in 0.0067 secs.
  50% in 0.0076 secs.
  75% in 0.0081 secs.
1458011507    56.512µs      126
  90% in 0.0094 secs.
  95% in 0.0129 secs.
```

**Comment by gyuho** [comment]
@xiang90 I think it's just Google cloud console being mixed up in the output. If you run locally, you won't get it:

```
Latency distribution:
  10% in 0.0501 secs.
  25% in 0.0742 secs.
  50% in 0.1367 secs.
  75% in 0.1925 secs.
  90% in 0.2330 secs.
  95% in 0.2746 secs.
  99% in 0.3204 secs.
Sample in one second (unix latency throughput):
1458012538    68.818µs        1
1458012539    57.151µs        7
1458012540    64.563µs       10
1458012541    52.074µs        8
1458012542    49.015µs        6
1458012543    60.327µs        7
1458012544    51.645µs        5
1458012545    63.566µs        8
1458012546    49.346µs        6
1458012547    60.15µs        6
1458012548    57.226µs        6
1458012549    57.738µs        8
1458012550    51.174µs       10
1458012551    61.703µs        6
1458012552    61.581µs        6

```

Thanks.

**Comment by xiang90** [comment]
Also the latency does not make sense to me. If the ave latency is 50us, the throughput should be 1second/50us = 20 000

**Comment by gyuho** [comment]
@xiang90 i will do more testing today to double check on this. Thanks!

**Comment by gyuho** [comment]
@xiang90 We sum the actual latency per request. So during 1 second:

```
req #1: takes 10ns
req #2: takes 30ns
```

Then total request count 2 and average latency is 20ns, not (1s/2=500ms).
Did you mean this?

File: tools/benchmark/cmd/report.go [hunk 7]
```diff
@@ -85,6 +90,9 @@ func (r *report) finalize() {
 		if res.errStr != "" {
 			r.errorDist[res.errStr]++
 		} else {
+			now := time.Now()
+			actualLat := now.Sub(res.happened)
```
Comment (xiang90) [inline tools/benchmark/cmd/report.go]: why actualLat = now.Sub(res.happened)?

File: tools/benchmark/cmd/report.go [hunk 8]
```diff
@@ -85,6 +90,9 @@ func (r *report) finalize() {
...
+			r.sps.Add(res.happened, actualLat)
```
Comment (xiang90) [inline tools/benchmark/cmd/report.go]: i believe this should be (res.happened, res.duration)

Comment (gyuho) [inline tools/benchmark/cmd/report.go]: You are right. Will fix and rerun the benchmark.

**Comment by gyuho** [comment]
@xiang90 Just fixed. Another run:

```

./benchmark --endpoints="10.240.0.9:2378,10.240.0.10:2378,10.240.0.11:2378" \
	put --sample --conns=1 --clients=1 --total=5000 --val-size=3 --key-space-size=1

Summary:
  Total:        40.0143 secs.
  Slowest:      0.0241 secs.
  Fastest:      0.0046 secs.
  Average:      0.0080 secs.
  Stddev:       0.0022 secs.
  Requests/sec: 124.9552
Response time histogram:
  0.005 [1]     |
  0.007 [833]   |∎∎∎∎∎∎∎∎∎
  0.008 [3389]  |∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
  0.010 [358]   |∎∎∎∎
  0.012 [143]   |∎
  0.014 [101]   |∎
  0.016 [72]    |
  0.018 [55]    |
  0.020 [32]    |
  0.022 [9]     |
  0.024 [7]     |
Latency distribution:
  10% in 0.0063 secs.
  25% in 0.0068 secs.
  50% in 0.0077 secs.
  75% in 0.0081 secs.
1458056008   8.034968ms     125
  90% in 0.0095 secs.
  95% in 0.0130 secs.
  99% in 0.0181 secs.
Sample in one second (unix latency throughput):
1458055996   7.895852ms     120
1458055997   8.163077ms     122
1458055998   8.096864ms     124
1458055999   8.017012ms     125
1458056000   7.732845ms     129
1458056001   7.997344ms     125
1458056002   8.175713ms     122
1458056003    8.16294ms     123
1458056004   8.187391ms     122
1458056005   7.919586ms     126
1458056006   7.960912ms     125
1458056007   7.949822ms     126
1458056008   8.034968ms     125
1458056009     8.0809ms     123
1458056010   8.001885ms     125
1458056011   7.732412ms     129
1458056012   7.745179ms     130
1458056013   8.047295ms     124
1458056014   8.058717ms     124
1458056015   8.000064ms     125
1458056016   8.044817ms     124
1458056017   8.004889ms     125
1458056018   7.887517ms     127
1458056019   7.961879ms     125
1458056020   7.870518ms     127
1458056021   7.970397ms     126
1458056022   7.655828ms     130
1458056023   8.139674ms     123
1458056024   8.097099ms     123
1458056025   7.960426ms     126
1458056026   8.308047ms     120
1458056027   8.063992ms     123
1458056028    7.93717ms     127
1458056029   8.079418ms     123
1458056030   8.066746ms     125
1458056031   8.226309ms     121
1458056032   7.973733ms     126
1458056033   8.038257ms     124
1458056034   7.897846ms     127
1458056035   7.907409ms     126
1458056036   8.400641ms       8





./benchmark --endpoints="10.240.0.9:2378,10.240.0.10:2378,10.240.0.11:2378" \
	put --sample --conns=5 --clients=2000 --total=100000 --val-size=3 --key-space-size=1
Summary:
  Total:        3.3786 secs.
  Slowest:      0.1333 secs.
  Fastest:      0.0152 secs.
  Average:      0.0668 secs.
  Stddev:       0.0201 secs.
  Requests/sec: 29598.0978
Response time histogram:
  0.015 [1]     |
  0.027 [104]   |
  0.039 [4993]  |∎∎∎∎∎∎∎
  0.051 [17237] |∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
  0.062 [25961] |∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
  0.074 [19782] |∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
  0.086 [14186] |∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
  0.098 [8469]  |∎∎∎∎∎∎∎∎∎∎∎∎∎
  0.110 [5651]  |∎∎∎∎∎∎∎∎
  0.121 [3387]  |∎∎∎∎∎
  0.133 [229]   |

Latency distribution:
  10% in 0.0419 secs.
  25% in 0.0521 secs.
  50% in 0.0633 secs.
  75% in 0.0803 secs.
  90% in 0.0969 secs.
  95% in 0.1050 secs.
  99% in 0.1161 secs.
Sample in one second (unix latency throughput):
1458056063   62.042974ms    6738
1458056064   63.623019ms   31156
1458056065   63.622436ms   31437
1458056066   74.144639ms   26869
1458056067   74.832193ms    3800
```

**Comment by xiang90** [comment]
lgtm

**Comment by gyuho** [comment]
thanks. will do more testings with this patch.

**Pull request merged**